### PR TITLE
The Great Fabric Unification

### PR DESCRIPTION
**Pull request by kosma**
Here it is - the thing I've been working on for a long long time. With each next Fabric port, new ports got love and changes, while old ports stayed there, old and sad, without new features and fixes. At the same time, Fabric's architecture inherently promises as much forward compatibility as possible - meaning it should be entirely possible to make a single mod with only a bit of version-specific shim logic to adapt for Vanilla/Yarn changes.

The had part was figuring out the correct Gradle magic. Roughly 50% of the magic comes from [ViaFabric](https://github.com/ViaVersion/ViaFabric) - with the rest I got help from modmuss50 (Fabric author). The end result uses native Fabric Loader functionality of child mods, and is very simple and elegant considering the magic it does.

TL;DR The Fabric JAR now contains a bunch of version-specific child mods, and Fabric Loader automatically chooses the matching one.

## The Process

1. Backport a whole bunch of changes down to 1.14/1.15. That includes things that don't exist in those versions, like negative Y handling. This is to minimize code differences between versions as much as possible.
2. Revert invasive changes that only existed in 1.18 - namely, my `onchunkgenerate` and `signmarkers` feature branches. Rationale here is that it's much easier to reapply those changesets after unification.
3. Add `FabricVersionAdapter` methods for all things that are known to be version-specific at that point, and refactor out all the version-specific calls into per-version adapter classes.
4. Create a separate subproject (initially `fabric-helper`, later changed to just `fabric`) and move all version-independent code there.
5. Add all the necessary `build.gradle` and `fabric.mod.json` logic.
6. Backport changes from `onchunkgenerate` and `signmarkers` feature branches.
7. Add `FabricVersionAdapter` methods that were not obvious previously (e.g. same field/method names but type/signature changes).

## Verification

The commit log is purposely very long - I spent an unreasonable amount of time splitting changes into logical chunks, to make sure I am not missing anything or changing something by accident. For verifying the end result (changes between "old" `fabric-1.18` mod and "new" `fabric` mod) I recommend running:

```bash
git clone https://github.com/webbukkit/dynmap ../dynmap-upstream
find ../dynmap-upstream/fabric-1.18/ -name '*.java' | xargs sed -e 's/org.dynmap.fabric_1.18/org.dynmap.fabric/' -i
diff -urw ../dynmap-upstream/fabric-1.18/src/main/java/org/dynmap/fabric_1_18/ fabric/src/main/java/org/dynmap/fabric/
```

## TODO

- ~~Fix import ordering (improves parity with old mod + makes things nicer)~~
- ~~Backport cleanup commits (some stuff is still showing up on diffs)~~
- ~~Write architecture documentation~~
- ~~Write porting documentation~~
- ~~Check existing mixins for possible deduplication opportunities~~
- ~~Write version adapters where needed~~
- ~~Double-check the FabricMapChunkCache changes, possibly simplify even more~~
- ~~Verify all functionalities on all versions~~
- ~~Remove temporary commits~~
- ~~Sign off on the pull request~~

Once this is complete, I can start the actual work I've been meaning to do - cleaning up the mod code, fixing some bugs, and adding functionalities where missing.

### CODE REVIEW

**Comment by kosma** [comment]
Ready for review. I have no idea if ready for merging - I've been working on this for such a long time that I need more eyes to look and figure out if anything is missing or wrong.

**Comment by mikeprimm** [comment]
Excellent - I'll give it a look tonight, and we'll probably merge it and sort through anything that pops out :)

**Comment by mikeprimm** [comment]
This is a big one, but with how much we've talked about this as you developed it, I'm good with us giving this a go and hammering out anything that might need fixing.  Thanks!!
